### ISSUE TITLE
C++ async API does not provide a way to wake up an event loop

### ISSUE THREAD

**Issue by andrewrk** [grpc/grpc#20194]
Re-opening #14796 because stale bot closed it.

In order for an API to integrate into a foreign event loop, it needs to provide a polling interface as well as a callback function which can be used to signal a mutex condition to wake up, or activate an eventfd, or put an item on a kqueue, etc.

`CompletionQueue` has `AsyncNext` which provides the polling interface, but it does not provide a way to wake up an event loop, or otherwise run a callback when the queue becomes non-empty.

Because of this missing feature gRPC is useless in an event-driven C++ codebase.

**Comment by markdroth** [comment]
We're ultimately going to address this by moving to EventManager-based polling, but there's no ETA yet.  Keeping this open for tracking purposes.

**Comment by andrewrk** [comment]
FWIW I'm personally no longer subscribed to this issue. Let's see if [Zig](https://ziglang.org/) gains support for event-based I/O grpc before this repo does  :smile:

**Comment by RobertBColton** [comment]
We're trying to get this with our Qt project right now and are going to use the thread solution in the interim. However, I do have great news for people interested in this with Qt applications.

The Qt main event loop uses the Win32 function MsgWaitForMultipleObjectsEx to determine when to wake up from keyboard input, mouse input, native timers, and other events. Such a callback only needs to signal an event on the handle Qt uses internally in the main event loop to wake up a Qt application when a GRPC event occurs.

**Comment by veblush** [comment]
@RobertBColton @andrewrk Can you elaborate the requirement and use-case more? We're revisiting this issue again and it doesn't seem 100% clear to me why this is necessary for your cases and what gRPC API should look like if it's present. Since   the IO manager does a lot of things, customizing entire of it would be overkill if you just want to wake your event-loop if there is an event. If you have a single thread to process all incoming events, maybe just having one-callback function which will be called for every gRPC message incoming event and misc other events such as on_connect, on error, and on_close might be enough and ideal.

**Comment by andrewrk** [comment]
> FWIW I'm personally no longer subscribed to this issue. 

plz no ping

**Comment by RobertBColton** [comment]
@markdroth @andrewrk 
Is this not a duplicate of #14565 or vice versa? I don't think we really need both issues open for people to follow updates on this.

@veblush
I want to have a single thread, and to run both a GUI event loop and a GRPC event loop, concurrently.

**Comment by markdroth** [comment]
@RobertBColton Thanks for pointing that out.  I've closed the other issue in favor of this one.

**Comment by hkadayam** [comment]
We have similar requirement as well, where we wanted to run GRPC in the same thread as the one doing IO and wanted to multiplex both network and IO. Also we wanted to integrate with our custom event loop which does ePoll as well as tight loop polling (Intel SPDK). Without this we are down to either faking system call like ePoll and capturing that or dedicate a thread to do grpc handling and other threads to do IO - both not ideal. Thanks!!

**Comment by ledergec** [comment]
Trying to explain why integrating into an existing event-loop may be interesting from my experience.

In the past two larger projects I have been working on we decided to use the following design:
- we used multiple micro services
- every service was single threaded and driven by an event-loop

Being Single Threaded had two effects:
1. no multi threading bugs
2. it is possible to simulate time

The latter point was a success. We were able to record received data let's say at 1Hz and replay it fully deterministic with correct timestamps etc. at the maximum speed which was permitted by the CPU. This also holds true for interleaved data from multiple IOs. So, many of our higher level tests could be simply recorded data replayed - and they would still execute fast.

**Comment by bsergean** [comment]
We would also like to have this supported, so that we don't have to create new threads ; in our case we'd like to use grpc as a client, and the only solution I see is to create a background thread that does the grpc work, and send event to it.

Our event loop is libuv, but once this is done for one loop it should work for all or most of them.

libcurl supports an external event loop with it multi-interface. 

* https://curl.se/libcurl/c/libcurl-multi.html
* https://curl.se/libcurl/c/multi-uv.html / this is the interesting example, it shows the integration with libuv
* https://curl.se/libcurl/c/multi-event.html / same one, but with libevent.

libcurl still manage the connections, and can be hooked to many external libraries so it would be a good thing to look at.

It requires one timer (most libraries should have this), and one way to attach a file descriptor managed by the library (grpc), so that it can be polled by the 'evented' library.

Apparently libwebsocket also has this
* https://libwebsockets.org/git/libwebsockets/tree/lib/event-libs

**Comment by drfloob** [comment]
We're working on it! https://github.com/grpc/grpc/blob/master/include/grpc/event_engine/event_engine.h.

It's a complicated thing to land cleanly on every platform that gRPC supports. We expect the work to be completed in 2022, with the first internal usage coming this month.

**Comment by bsergean** [comment]
Great to hear that this is actively being worked on ! I imagine it's quite complicated to implement well. 

If there's a branch or something available for testing we'll (try) to give it a try.

**Comment by yaram** [comment]
Has any progress been made on this since 2022?

**Comment by rockwotj** [comment]
There is an (hacky IMO) way to do this with grpc::Alarm in the C++ SDK

https://github.com/grpc/grpc/blob/1309eb283c3e11c471191f286ceab01b75477ffc/include/grpcpp/alarm.h#L71

**Comment by mainak-agrawal** [comment]
https://grpc.github.io/grpc/core/namespacegrpc__event__engine_1_1experimental.html

I see that the above is categorized as experimental. Is there any timeline available for when it would be available for production use please?
